### ISSUE TITLE
comments on Linux audio core

### ISSUE THREAD

**Issue by davephillips** [VCVRack/Rack#407]
Hi Andrew,

Apologies if this isn't where to post these comments, I'm not sure where else to post them. The new Linux audio core is problematic in a couple of ways.

My system is set up for low-latency with Jack. PortAudio performed well with it, giving me a clean interface for I/O connections in QJackCtl. I do have the PulseAudio Jack Sink installed, which is normally utilized for YouTube, Netflix, etc., and not for anything like serious audio work. The problem with Pulse is that it gives me choices only for Input or Output, not both.

Direct ALSA connections via the system interface restores full duplex I/O but has its own problems  (see my earlier report). Btw, at the close of almost every session Rack reports a segfault, e.g. :

Makefile:50: recipe for target 'run' failed
make: *** [run] Segmentation fault (core dumped)

So I'm wondering if there is a way to build Rack without the pulse-simple stuff. Closer to direct Jack support is better on a system like mine.

dp

**Comment by MarcBoule** [comment]
Hi,

I'm also using a low latency kernel, in Ubuntu 17.10, and the latest Rack dev build, with _PulseAudio Jack Sink_, and I'm not getting the segfault. I'm available for some testing if you guys need it.

I was also wondering if Jack will be supported directly in the devices of the audio interface in Rack. Since the move from portaudio to rtaudio, it no longer shows up in the list.

Cheers,

Marc

**Comment by davephillips** [comment]
I've got a little further. When I select the device labeled 'system 17 in, 15 out' it opens an RtApiJack connection with 8 ports I/O, matching the configuration of the Audio Interface module. So far things are running without troubles but I'm testing only very light loads for the moment.

A screenshot is at http://linux-sound.org/images/whatisee-audiodevices.png.

Marc, does your device selection look anything like that ?

Strangely (?), when that system interface is selected, another entry appears in the Devices menu, for  'RtApiJack 8 in, 8 out' but if I select it the connection closes and the menu item disappears. So, it seems that I'm good as long as the RtApiJack interface actually opens when I select the system device. Btw, this message shows up after the plugins are loaded :

[warning] Failed to open audio stream: RtApiJack::probeDeviceOpen: the requested sample rate (44100) is different than the JACK server rate (48000).

But I'm not sure it matters. For now I'm not getting a segfault on every Quit and the RtApiJack connection does seem to work okay.

**Comment by AndrewBelt** [comment]
Run this test https://github.com/AndrewBelt/rtaudio_test

Post the terminal output, as well as a list of which device numbers correctly output audio.

**Comment by MarcBoule** [comment]
Dave: my jack connections look similar to yours, but without RtApiJack. So I have two devices on each side, with two connections: jack sink <-> system and system <-> jack source. In Rack, the device I select is called "Default (32 in, 32 out)".
Andrew: here's the test on my end, in case it can be useful for you (output below), but since I'm not seeing any RtApiJack anywhere, it's probably not relevant.

"
[debug] deviceCount: 6

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.

[debug] Stream 0: probed: 0, name:  outputs: 0, inputs: 0, duplex: 0
[debug] RtApi::openStream: input and output StreamParameters structures are both NULL!

terminate called after throwing an instance of 'RtAudioError'
  what():  RtApi:: a stream is not open!
Aborted (core dumped)
make: *** [run] Error 134
"

By the way, when I start Rack dev I get a bunch of messages like these:
RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.
RtApiAlsa::probeDeviceOpen: unable to synchronize input and output devices.

Rack dev is working fine though, with the jack connections and device described above in my comment.

**Comment by AndrewBelt** [comment]
Well, we've ruled out Rack, so the problem is with your OS and RtAudio.

I've modified the rtaudio_test script to handle devices with 0 outputs and 0 inputs (I don't know what that even exists), so at least it will get past your first audio device. Pull and try it again.

**Comment by MarcBoule** [comment]
Here's the new output on my end. 
Stream 1 produces sound on my laptop speakers
Stream 5 produces sound through my USB audio interface (where I send Rack and other DAWs).

"
LD_LIBRARY_PATH=../dep/lib ./rtaudio_test
[debug] deviceCount: 6

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.

[debug] Stream 0: probed: 0, name:  outputs: 0, inputs: 0, duplex: 0
[debug] Stream 1: probed: 1, name: hw:HDA Intel PCH,0 outputs: 4, inputs: 2, duplex: 2

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.

[debug] 	outputs: 1, inputs: 4, blockSize: 2, sampleRate: 256
[debug] Stream 2: probed: 1, name: hw:HDA Intel PCH,3 outputs: 8, inputs: 0, duplex: 0

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.

[debug] 	outputs: 2, inputs: 8, blockSize: 0, sampleRate: 256
[debug] Stream 3: probed: 1, name: hw:HDA Intel PCH,7 outputs: 8, inputs: 0, duplex: 0

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.

[debug] 	outputs: 3, inputs: 8, blockSize: 0, sampleRate: 256
[debug] Stream 4: probed: 1, name: hw:HDA Intel PCH,8 outputs: 8, inputs: 0, duplex: 0

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.

[debug] 	outputs: 4, inputs: 8, blockSize: 0, sampleRate: 256
[debug] Stream 5: probed: 1, name: default outputs: 32, inputs: 32, duplex: 32

RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.


RtApiAlsa::getDeviceInfo: snd_pcm_open error for device (hw:0,0), Device or resource busy.


RtApiAlsa::probeDeviceOpen: unable to synchronize input and output devices.

[debug] 	outputs: 5, inputs: 8, blockSize: 8, sampleRate: 256
"

**Comment by AndrewBelt** [comment]
Surround code samples by ``` by the way.

Does this match the result you're getting in Rack? The numbers should correspond to the order in the Audio Interface dropdown.

**Comment by davephillips** [comment]
The output from here:

[dlphilp@The6300 rtaudio_test]$ make run
LD_LIBRARY_PATH=../dep/lib ./rtaudio_test
[debug] deviceCount: 3
[debug] Stream 0: probed: 1, name: system outputs: 15, inputs: 17, duplex: 15
[debug] RtApiJack::probeDeviceOpen: the requested sample rate (44100) is different than the JACK server rate (48000).

terminate called after throwing an instance of 'RtAudioError'
  what():  RtApi:: a stream is not open!
Makefile:15: recipe for target 'run' failed
make: *** [run] Aborted (core dumped)

It's not switching the sample rate ? I run Jack at 48k all the time, btw.

**Comment by AndrewBelt** [comment]
Oh, interesting. I'm requesting 44.1k in rtaudio_test, but it looks like it outright *fails* if I don't guess the right one.

Let me patch Rack and I'll post back.

**Comment by MarcBoule** [comment]
The 6 devices do show up, and in the same order
Here's exactly what I see in Rack's Audio interface:

No device
(0 in, 0 out)
HDA Intel PCH,0 (2 in, 4 out)
HDA Intel PCH,3 (0 in, 8 out)
HDA Intel PCH,7 (0 in, 8 out)
HDA Intel PCH,8 (0 in, 8 out)
default (32 in, 32 out)

In Rack I get sound with "default", but not with PCH,0

**Comment by AndrewBelt** [comment]
For the Jack sample rate mismatch problem, does setting the sample rate *before* setting the device work?

**Comment by davephillips** [comment]
Hi Andrew, If I understand the question, yes, the interface behaves correctly in that regard, i.e. Jack is set for 48k, so the interface disallows any device that isn't set at 48k.

**Comment by davephillips** [comment]
I altered the sr in rtaudio_test.cpp to 48000 and ran the test again. This time I got a better report. Btw, audio output was heard only on the first device, i.e. the system device. Is that expected ?

[dlphilp@The6300 rtaudio_test]$ make run
LD_LIBRARY_PATH=../dep/lib ./rtaudio_test
[debug] deviceCount: 3
[debug] Stream 0: probed: 1, name: system outputs: 15, inputs: 17, duplex: 15
[debug] 	outputs: 0, inputs: 8, blockSize: 8, sampleRate: 128
[debug] Stream 1: probed: 1, name: PulseAudio JACK Sink outputs: 0, inputs: 10, duplex: 0
[debug] 	outputs: 1, inputs: 0, blockSize: 8, sampleRate: 128
[debug] Stream 2: probed: 1, name: PulseAudio JACK Source outputs: 12, inputs: 0, duplex: 0
[debug] 	outputs: 2, inputs: 8, blockSize: 0, sampleRate: 128
